### PR TITLE
Allow plugins to add additional work package tabs

### PR DESCRIPTION
**Pull request by tessi**
This is the part that prepares the OP core for [a better GitHub integration](https://community.openproject.com/projects/openproject/work_packages/36132/activity). Specifically, this PR adds additional APIs to OpenProject core so that plugins can display additional work package tabs.

## State of the PR 🏗️ 

* [x] Tests needs to be written
* [x] Make linters go ✅ 
* [x] I want to confirm with @ulferts or @oliverguenther whether the introduced functionality is "in the spirit of OP core"
* [x] I implemented additional tabs for work packages for the [`WorkPackagesFullViewComponent`](https://github.com/tessi/openproject/blob/876d48057fe7779fd7849928d8f71ed181a25671/frontend/src/app/modules/work_packages/routing/wp-full-view/wp-full-view.component.ts#L49-L49), do I need to also implement `wp-split-view`? If so, how can I reach this split-view?
* [x] I commented on this PR with a couple of questions which needs answers/confirmation
* [x] [CLA](https://docs.openproject.org/development/#openproject-contributor-license-agreement-cla)

## Feature Overview 📑 

https://user-images.githubusercontent.com/206108/107437569-7d044980-6b2f-11eb-8e8f-00c01925c7fc.mov

* It's possible for a plugin to register a a component as an additional work package details-view tab in a type-safe way
  ```ts
  function displayable(work_package: WorkPackageResource): boolean {
    return(!!work_package.github);
  }
  
  export function initializeGithubIntegrationPlugin(injector:Injector) {
    const hooks = injector.get<HookService>(HookService);
    hooks.registerWorkPackageTab(
      new Tab(
        GitHubTabComponent,
        I18n.t('js.github_integration.work_packages.tab_name'),
        "github",
        displayable
      )
    );
  }
  ```
* it's possible for the ruby-side of a plugin to define permissions for additional tabs by extending `API::V3::WorkPackages::WorkPackageRepresenter#additional_work_package_tabs` e.g. in a patch like this:
  ```ruby
  module OpenProject::GithubIntegration::Patches
    module Api
      module V3
        module WorkPackageRepresenter
          module_function

          def extension
            ->(*) do
              link :github,
                  uncacheable: true do
                next unless represented.project.module_enabled?(:github) && current_user.allowed_to?(:show_github_content, represented.project)

                {
                  href: "#{work_package_path(id: represented.id)}/additionalTabs/github",
                  title: "github"
                }
              end
            end
          end
        end
      end
    end
  end
  ```
* works for both work package views (split and full)

A demo of how the GitHub integration would use that new API can be found [at the sister-PR](https://github.com/tessi/openproject/pull/1).

Note: This is a re-opening of #8988 but not from my private op-fork - this should convince github actions to run the tests.

work package: https://community.openproject.org/projects/openproject/work_packages/36185

### CODE REVIEW

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<div class="detail-panel-description detail-panel--integration detail-panel--autocomplete-target"
+     *ngIf="workPackage">
+  <div class="detail-panel-description-content">
+    <ng-container #tabContent></ng-container>
```
Comment (tessi) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: note: @oliverguenther and me wanted to enhance this part in a pairing session ℹ️  (could be in a separate PR though)

**Comment by tessi** [comment]
From what I see we currently don't really unit-tests components, but feature-spec them. I like that idea, but it also means there is not much to feature-spec for this PR (as visible changes will only come when doing the actual github integration).

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html (hunk 1 above)
Comment (oliverguenther) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: Use something like this in the template instead of the whole `ngAfterViewInit`.

```
<ndc-dynamic
  *ngIf="tab"
  [ndcDynamicComponent]="tab.component"
  [ndcDynamicInputs]="{ workPackage: workPackage, tab: tab }"
  >
</ndc-dynamic>
```

Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: cool, thanks. I see we already use ndc-dynamic in other places 🎉  will do 🐜

File: frontend/src/app/modules/hal/resources/work-package-resource.ts [hunk 2]
```diff
@@ -182,6 +183,10 @@ export class WorkPackageBaseResource extends HalResource {
     return otherWorkPackage.parent?.$links.self.$link.href === this.$links.self.$link.href;
   }
 
+  public tabs(registeredTabs: Tab[]):Tab[] {
+    return _.filter(registeredTabs, (tab) => tab.displayable(this));
```
Comment (oliverguenther) [inline frontend/src/app/modules/hal/resources/work-package-resource.ts]: I'm not too happy with the `tabs` being added to the resource itself. It doesn't really have anything to do with the work package resource itself. It could just be a private member of the `TabWrapperComponent` ?

**Comment by oliverguenther** [comment]
> From what I see we currently don't really unit-tests components, but feature-spec them. I like that idea, but it also means there is not much to feature-spec for this PR (as visible changes will only come when doing the actual github integration).

Yes that's correct. You could of course test the tab wrapper component, but there is not much use in it. It's fine for me to delay the feature test to the github integration.

Great stuff Tessi :clap: We'll need to hold this PR a bit longer until a `release/11.2` branch is opened and we can merge it for 11.3. I'll also want to look into the way the current default tabs work, as ideally we'd move them into the same behavior. There's still the a11y issue to be resolved as well correct?

**Comment by tessi** [comment]
>There's still the a11y issue to be resolved as well correct?

Yes. As far as I remember, we wanted to solve that together with moving the defaults tabs 🤔

File: frontend/src/app/modules/hal/resources/work-package-resource.ts (hunk 2 above)
Comment (tessi) [reply frontend/src/app/modules/hal/resources/work-package-resource.ts]: yep, it's using the `workPackage` resource as kind of a helper module. I think, since the implementation became quite simple with the recent refactorings, that we can just inline the implementation.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<div class="detail-panel-description detail-panel--integration detail-panel--autocomplete-target"
```
Comment (ulferts) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: Are those classes necessary?

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    const registeredTabs = this.hooks.getWorkPackageTabs();
+    const displayableTabs = _.filter(
+      registeredTabs,
+      (tab) => tab.displayable(this.workPackage)
+    );
+    this.tab = _.find(displayableTabs, ({identifier: id}) => id === tabIdentifier);
+  }
+
+  ngOnInit() {
+    const wpId = this.workPackageId || this.$transition.params('to').workPackageId;
```
Comment (ulferts) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: I suggest extracting this into a method, but that might be personal preference

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import {Component, ComponentFactoryResolver, Input, OnInit} from '@angular/core';
+import {I18nService} from 'core-app/modules/common/i18n/i18n.service';
+import {WorkPackageResource} from 'core-app/modules/hal/resources/work-package-resource';
+import {HookService} from 'core-app/modules/plugins/hook-service';
+import {UntilDestroyedMixin} from 'core-app/helpers/angular/until-destroyed.mixin';
+import {APIV3Service} from 'core-app/modules/apiv3/api-v3.service';
+import { Tab } from './tab';
+
+@Component({
+  templateUrl: './tab-wrapper.html',
```
Comment (ulferts) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: I believe (and here others might know better) that the template should be named similar to how the component is named 

/cc @oliverguenther, @aleixsuau

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  selector: 'wp-tab',
+})
+export class WorkPackageTabComponent extends UntilDestroyedMixin implements OnInit {
+  @Input() public workPackageId?:string;
```
Comment (ulferts) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: Does it have to be public?

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    this
+      .apiV3Service
+      .work_packages
+      .id(wpId)
+      .requireAndStream()
+      .pipe(
+        this.untilDestroyed()
+      )
+      .subscribe((wp) => {
+        this.workPackageId = wp.id!;
```
Comment (ulferts) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: What do you need to update the variable for? The id is static so it will not change between calls.

At the latest, if you extract the method as suggested above you should no longer need it.

File: frontend/src/app/modules/plugins/hook-service.spec.ts [hunk 8]
```diff
@@ -156,4 +160,31 @@ describe('HookService', function() {
...
+    })
+    class MyTestTabComponent implements TabComponent {
+      workPackage: WorkPackageResource;
+    }
+
+    const tab1 = new Tab(MyTestTabComponent, 'GitHub', 'github', () => true);
+    const tab2 = new Tab(MyTestTabComponent, 'GitLab', 'gitlab', () => true);
+    const tab3 = new Tab(MyTestTabComponent, 'Bitbucket', 'bitbucket', () => true);
+
+    it('returns empty results', function() {
```
Comment (ulferts) [inline frontend/src/app/modules/plugins/hook-service.spec.ts]: You are lying here! 🙈

**Review by ulferts — DISMISSED** [review]
Overall, it looks solid to me. What I like about it especially, is that we should be able to reuse the same functionality for the 'default' tabs (activity, relations, watchers).

However, it would be good if somebody more versed in the front end (e.g. @aleixsuau or @b12f) could take a second look

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 5 above)
Comment (aleixsuau) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: Right, tab-wrapper.component.html. You can use the [Angular CLI](https://angular.io/cli/generate) to generate components (ng generate component componentName (ng g c )) and it will generate the component files properly.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 6 above)
Comment (aleixsuau) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: Properties are public by default so there is no need to set them explicitly.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  findTab() {
+    const tabIdentifier = this.$transition.params('to').tabIdentifier;
+    const registeredTabs = this.hooks.getWorkPackageTabs();
+    const displayableTabs = _.filter(
+      registeredTabs,
+      (tab) => tab.displayable(this.workPackage)
+    );
+    this.tab = _.find(displayableTabs, ({identifier: id}) => id === tabIdentifier);
+  }
```
Comment (aleixsuau) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: Since registeredTabs is an array of tabs, I think it is better to use the native find and filter array methods. This way probably we will be able to reduce unnecessary third-party libraries and kbs in the future.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  @Input() public workPackageId?:string;
+
+  public tab:Tab|undefined;
+  public workPackage:WorkPackageResource;
+
+  public constructor(readonly I18n:I18nService,
+                     readonly $transition:Transition,
+                     readonly apiV3Service:APIV3Service,
+                     readonly hooks:HookService,
+                     private componentFactoryResolver: ComponentFactoryResolver) {
```
Comment (aleixsuau) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: It seems that componentFactoryResolver is not used.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab.component.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { Component } from '@angular/core';
+import { WorkPackageResource } from 'core-app/modules/hal/resources/work-package-resource';
+
+export interface TabComponent extends Component {
```
Comment (aleixsuau) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab.component.ts:4]: The standard approach is to create components by using the @Component decorator.

File: frontend/src/app/modules/hal/resources/work-package-resource.ts [hunk 12]
```diff
@@ -46,6 +46,7 @@ import {WorkPackagesActivityService} from "core-components/wp-single-view-tabs/a
 import {WorkPackageNotificationService} from "core-app/modules/work_packages/notifications/work-package-notification.service";
 import {InjectField} from "core-app/helpers/angular/inject-field.decorator";
 import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
+import { Tab } from 'core-app/components/wp-single-view-tabs/tab/tab';
```
Comment (aleixsuau) [inline frontend/src/app/modules/hal/resources/work-package-resource.ts]: It seems that this is not used.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import {HookService} from 'core-app/modules/plugins/hook-service';
+import {UntilDestroyedMixin} from 'core-app/helpers/angular/until-destroyed.mixin';
+import {APIV3Service} from 'core-app/modules/apiv3/api-v3.service';
+import { Tab } from './tab';
+
+@Component({
+  templateUrl: './tab-wrapper.html',
+  selector: 'wp-tab',
+})
+export class WorkPackageTabComponent extends UntilDestroyedMixin implements OnInit {
```
Comment (aleixsuau) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: The file name and class name are usually the same.

File: frontend/src/app/components/wp-single-view-tabs/tab/tab.component.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+import { Component } from '@angular/core';
```
Comment (aleixsuau) [inline frontend/src/app/components/wp-single-view-tabs/tab/tab.component.ts]: You could place interfaces into a typings.d.ts (existing or a new one in the feature's module folder) and it will be available globally without having to import it. Since they are "compile code" it doesn't make sense to include them in the final bundle.

**Review by aleixsuau — APPROVED** [review]
Hi @tessi ,

Thanks, it is great to have dynamic plugin tabs in OP!

I'd suggest the following minor changes:

- Name the files something like "plugin-tab" in order to differentiate them from other tab functionalities.
- Encapsulate the files in a module (e.g. PluginTabsModule). It would be great if the module would include a little documentation about how it works and why (/*In [Typedoc](https://typedoc.org/guides/doccomments/) comment format*/).
- Encapsulate the feature in a component (e.g. PluginTabsComponent) that will get the tabs (hooks.getWorkPackageTabs()) and will display them (<li *ngFor="let tab of tabs()"...). This way we'd able to use it directly (<op-plugin-tabs [wpId]="2">) everywhere (wp-full-view.html, wp-split-view.html, maybe others in the future).

Best

**Comment by tessi** [comment]
perfect, many thanks for you reviews. will address them today 💚

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html (hunk 3 above)
Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: honestly, I have to try. I copied them over from another tab and thought that classes should be consistent between tabs. but I can gradually remove them and see what breaks :D

Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.html]: seems these classes can go, and that let's us remove one levels of div-nesting 🎉

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 5 above)
Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: thanks for these hints, I have to re-learn angular with this project 💡

Comment (aleixsuau) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: You are very welcome, let me know if you need help whit that.

File: frontend/src/app/modules/plugins/hook-service.spec.ts (hunk 8 above)
Comment (tessi) [reply frontend/src/app/modules/plugins/hook-service.spec.ts]: ha, you found out 🙈 thanks for noticing, I would like to blame the copy&paste gods ;)

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 9 above)
Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: i agree to the fullest 💚

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 10 above)
Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: yes, that is a sad leftover (same as other unresolved dependencies). sorry you had to play the linter. but then, why didn't the actual linter catch this? shouldn't we automate this?

Comment (tessi) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: (I removed the unused variable, but leave this thread open to talk about the linter topic)

**Comment by tessi** [comment]
@aleixsuau many thanks for the review especially since you covered everything from low altitude (linting) to high altitude (architecture/naming) topics 🚀 

Your review comment suggests that this feature will be for plugin tabs only.

> Name the files something like "plugin-tab" in order to differentiate them from other tab functionalities.

As far as I understood @ulferts @oliverguenther and me wanted to use this new tab-system for all work package tabs eventually. So I will skip the suggested rename.

However, I will attempt your other suggestions (wrapping the feature in another component and making it a documented module)

**Comment by tessi** [comment]
Alright, I got most of your awesome feedback implemented, there is a catch however.
I had trouble implementing @aleixsuau 's suggestions:

> Encapsulate the files in a module (e.g. PluginTabsModule). It would be great if the module would include a little documentation about how it works and why (/In Typedoc comment format/).
> Encapsulate the feature in a component (e.g. PluginTabsComponent) that will get the tabs (hooks.getWorkPackageTabs()) and will display them (<li *ngFor="let tab of tabs()"...). This way we'd able to use it directly (<op-plugin-tabs [wpId]="2">) everywhere (wp-full-view.html, wp-split-view.html, maybe others in the future).

I find the theory very compelling, but by my life couldn't figure out how to extract all this into a new module and still have the paths behave properly for split-view and full-view. similar problems are the html templates, as they are a little different for both types of work package views.

To me it sounds like we should address this feedback once we migrate all tabs over to this new system. This would enable us to extract the full tabs-row (and content) instead of just parts of it.

Alternatively, we could do a pairing session. Maybe most of my troubles are easy fixes for you (e.g. when no error is thrown, everything compiled, but still components do not render. How could anyone debug these situations?)

File: frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts (hunk 10 above)
Comment (aleixsuau) [reply frontend/src/app/components/wp-single-view-tabs/tab/tab-wrapper.component.ts]: Hi @tessi ,
Not sure, the linter is working properly on my RubyMine. Are you using RubyMine? I had to switch to it from VSCode due to some issues, this project seemed to be too resource-demanding for it.

**Comment by aleixsuau** [comment]
> @aleixsuau many thanks for the review especially since you covered everything from low altitude (linting) to high altitude (architecture/naming) topics 🚀
> 
> Your review comment suggests that this feature will be for plugin tabs only.
> 
> > Name the files something like "plugin-tab" in order to differentiate them from other tab functionalities.
> 
> As far as I understood @ulferts @oliverguenther and me wanted to use this new tab-system for all work package tabs eventually. So I will skip the suggested rename.
> 
> However, I will attempt your other suggestions (wrapping the feature in another component and making it a documented module)

Sure, makes sense 👍

**Comment by aleixsuau** [comment]
> Alright, I got most of your awesome feedback implemented, there is a catch however.
> I had trouble implementing @aleixsuau 's suggestions:
> 
> > Encapsulate the files in a module (e.g. PluginTabsModule). It would be great if the module would include a little documentation about how it works and why (/In Typedoc comment format/).
> > Encapsulate the feature in a component (e.g. PluginTabsComponent) that will get the tabs (hooks.getWorkPackageTabs()) and will display them (<li *ngFor="let tab of tabs()"...). This way we'd able to use it directly (<op-plugin-tabs [wpId]="2">) everywhere (wp-full-view.html, wp-split-view.html, maybe others in the future).
> 
> I find the theory very compelling, but by my life couldn't figure out how to extract all this into a new module and still have the paths behave properly for split-view and full-view. similar problems are the html templates, as they are a little different for both types of work package views.
> 
> To me it sounds like we should address this feedback once we migrate all tabs over to this new system. This would enable us to extract the full tabs-row (and content) instead of just parts of it.
> 
> Alternatively, we could do a pairing session. Maybe most of my troubles are easy fixes for you (e.g. when no error is thrown, everything compiled, but still components do not render. How could anyone debug these situations?)

Hi @tessi ,
Of course, let's have a pairing session next week to see if the proposed solution can be done/makes sense. I'll contact you on Slack.

File: frontend/src/app/modules/work_packages/routing/wp-split-view/wp-split-view.html [hunk 15]
```diff
@@ -24,11 +24,9 @@
           <a href="" [textContent]="text.tabs.watchers"></a>
           <wp-watchers-count [wpId]="workPackage.id"></wp-watchers-count>
         </li>
-        <li *ngFor="let tab of tabs()" uiSref=".tabs"
-                  [uiParams]="{workPackageId: workPackage.id, tabIdentifier: tab.identifier}"
-                  uiSrefActive="selected">
-          <a href="" [textContent]="tab.displayName"></a>
-        </li>
+
+        <op-wp-tabs [workPackage]="workPackage" view="split"></op-wp-tabs>
```
Comment (tessi) [inline frontend/src/app/modules/work_packages/routing/wp-split-view/wp-split-view.html:28]: I definitely like this change ❤️

File: frontend/npm-shrinkwrap.json [hunk 16]
```diff
@@ -1,19303 +1,9 @@
 {
   "name": "openproject-frontend",
   "version": "0.1.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (tessi) [inline frontend/npm-shrinkwrap.json]: @oliverguenther what (apart from `npm install`) do I have to do to create something with `lockfileVersion` 2?

Comment (oliverguenther) [reply frontend/npm-shrinkwrap.json]: I have no clue why it would remove the shrinkwrap. Please just `git checkout origin/dev -- frontend/npm-shrinkwrap.json` as you have made no changes to the package.json. Usually, the shrinkwrap should auto-update once added dependencies.

**Review by oliverguenther — APPROVED** [review]
LGTM :+1:
